### ISSUE TITLE
Seeking Assistance for Making LightGBM Predictions in C# - Sample Code Needed

### ISSUE THREAD

**Issue by wil70** [microsoft/LightGBM#6113]
Hello,

I've successfully created a model using LightGBM CLI, specifically the C/C++ version. Now, I'm facing the task of performing predictions from a C# application. I've written the following code, but unfortunately, I'm encountering difficulties in getting it to work.

I've attempted various approaches, but having a simple example would be incredibly helpful.

Thank you in advance for your assistance.

Best regards,
Wil

```

using Microsoft.ML;
using System;
using System.Collections.Generic;
using System.Linq;

class ProgramWithMLNet
{
    static void Main()
    {
        var mlContext = new MLContext();

        // Load the trained model
        var model = mlContext.Model.Load(@"Path to a model that has been created with LightGBM CLI", out var modelSchema);

        // Prepare input data for prediction
        var input = new MyInputDataClass // Please replace this with your input data class
        {
            _data = new double[] { 0.1, .2, .3 }
        };

        // Make a prediction
        var predictionEngine = mlContext.Model.CreatePredictionEngine<MyInputDataClass, MyOutputDataClass>(model);
        var prediction = predictionEngine.Predict(input);

        // Display the prediction result
        Console.WriteLine("Prediction: " + prediction);
    }
}

class MyInputDataClass
{
    public double[] _data = null;
}

class MyOutputDataClass
{
    double _result = 0;
}
}
```

**Comment by guolinke** [comment]
you can refer to https://github.com/dotnet/machinelearning/tree/main/src/Microsoft.ML.LightGbm

**Comment by wil70** [comment]
Hello James,

**Regarding ML.Net,** I noticed that WrappedLightGbmBooster.cs includes EvalTrain(), which appears to take the entire dataset. However, my intention is to train with LightGBM CLI/C++ and then export the model for use in any language, such as C#. I'm considering whether I can call the following code directly:


```
double[] res = new double[1];
fixed (double* ptr = res)
    LightGbmInterfaceUtils.Check(WrappedLightGbmInterface.BoosterGetEval(Handle, dataIdx, ref outLen, ptr));
```
I expected  this to be a basic use case? if true, we should well-document it and it should be very easy to use.



**As for the 'task==convert_model,'** I'm simultaneously trying to use it in LightGBM, which generates the gbdt_prediction.cpp file. I've set up a small Visual Studio 2020 project to generate a DLL, with the hope of being able to call this DLL from C#. However, I'm encountering an issue with linking:

```

------ Build started: Project: LightGBM_TrainingResults_Cpp2CSharp, Configuration: Debug x64 ------
x64\Debug\gbdt_prediction.obj : warning LNK4042: object specified more than once; extras ignored
Creating library C:\Users\wilhe\source\repos\LightGBM_TrainingResults_Cpp2CSharp\x64\Debug\LightGBM_TrainingResults_Cpp2CSharp.lib and object C:\Users\wilhe\source\repos\LightGBM_TrainingResults_Cpp2CSharp\x64\Debug\LightGBM_TrainingResults_Cpp2CSharp.exp
gbdt_prediction.obj : error LNK2019: unresolved external symbol "public: void __cdecl LightGBM::Metadata::InitByReference(int,class LightGBM::Metadata const *)" (?InitByReference@Metadata@LightGBM@@QEAAXHPEBV12@@Z) referenced in function "public: void __cdecl LightGBM::Dataset::InitByReference(int,class LightGBM::Dataset const *)" (?InitByReference@Dataset@LightGBM@@QEAAXHPEBV12@@Z)
gbdt_prediction.obj : error LNK2019: unresolved external symbol "public: void __cdecl LightGBM::Metadata::Init(int,int,int,int,int)" (?Init@Metadata@LightGBM@@QEAAXHHHHH@Z) referenced in function "public: void __cdecl LightGBM::Dataset::InitStreaming(int,int,int,int,int,int,int)" (?InitStreaming@Dataset@LightGBM@@QEAAXHHHHHHH@Z)
C:\Users\wilhe\source\repos\LightGBM_TrainingResults_Cpp2CSharp\x64\Debug\LightGBM_TrainingResults_Cpp2CSharp.dll : fatal error LNK1120: 2 unresolved externals
Done building project "LightGBM_TrainingResults_Cpp2CSharp.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Build started at 9:18 PM and took 01.289 seconds ==========

```

Your assistance in resolving these issues would be greatly appreciated.

Thanks,

Wilhelm

**Comment by wil70** [comment]
Thanks, @guolinke!

I'm close to achieving my goal, which is to perform multiple predictions from a C# program using a model I've already trained with LightGBM CLI.

To clarify, I've exported the model with the 'output_model' parameter (default = LightGBM_model.txt, as described here: [LightGBM Parameters](https://lightgbm.readthedocs.io/en/latest/Parameters.html#output_model)).

In Python, I would do what I need like this:

```
import lightgbm as lgb

# Load the preprocessed data
data = ...

# Load the LightGBM model
model = lgb.Booster(model_file='lightgbm_model.txt')

# Make predictions
predictions = model.predict(data)

# Print the predictions
print(predictions)

```
How do I do this with c#?

The link you provided earlier explains how WrappedLightGbmTraining.Train(...) performs training, which I don't require. It returns a Booster that can then be used with WrappedLightGbmInterface.BoosterPredictForMat(...) for prediction. My question is, how can I obtain a Booster from an already saved model file (equivalent to the Python line: model = lgb.Booster(model_file='lightgbm_model.txt')) in C#?

Your guidance would be much appreciated.

Thanks,
Wil

**Comment by wil70** [comment]
This function inside WrappedLightGBMBooster is interresting, I might be able to load the model but to an InternalTreeEnsemble 
`
InternalTreeEnsemble GetModel(int[] categoricalFeatureBoundaries, string modelString)`

Then I need to figure out how to use the InternalTreeEnsemble ...to get a booster and it handle so I can call WrappedLightGbmInterface.BoosterPredictForMat(...)
A bit complicated but if this is the way to do it?

**Comment by wil70** [comment]
Hi @jameslamb, you're welcome to go ahead and close this thread.
I've experimented with a few solutions, and I've managed to make it work with LightGBM.Net (https://github.com/rca22/LightGBM.Net). While I don't have a fully operational and straightforward solution with ML.net, having a working option in LightGBM.Net suits my current needs. Hopefully, ML.net will eventually provide an out-of-the-box solution, which seems significant for Windows technology.

I appreciate your assistance.
Wil
